flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java

```diff
@@ -191,22 +191,28 @@ public boolean isEmpty() {
 	 */
 	public boolean prune(long pruningTimestamp) {
 		Iterator<Map.Entry<K, SharedBufferPage<K, V>>> iter = pages.entrySet().iterator();
-		boolean pruned = false;
+		List<SharedBufferEntry<K, V>> prunedEntries = new ArrayList<>();
 
 		while (iter.hasNext()) {
 			SharedBufferPage<K, V> page = iter.next().getValue();
 
-			if (page.prune(pruningTimestamp)) {
-				pruned = true;
-			}
+			page.prune(pruningTimestamp, prunedEntries);
 
 			if (page.isEmpty()) {
 				// delete page if it is empty
 				iter.remove();
 			}
 		}
 
-		return pruned;
+		if (!prunedEntries.isEmpty()) {
+			for (Map.Entry<K, SharedBufferPage<K, V>> entry : pages.entrySet()) {
+				entry.getValue().removeEdges(prunedEntries);
+			}
+			prunedEntries.clear();
```

Comment:
As prunedEntries is local now, you don't need to clear it any longer.

Reply:
Updated.